lib/api/apiUtils/rateLimit/tokenBucket.js

```diff
@@ -0,0 +1,291 @@
+/**
+ * WorkerTokenBucket - Local token buffer for rate limiting
+ *
+ * Each worker maintains a local token bucket that holds tokens granted by Redis.
+ * Tokens are consumed synchronously (in-memory) for each request, and refilled
+ * asynchronously in the background.
+ *
+ * This design keeps Redis OUT of the hot path:
+ * - Request handling: Pure in-memory token consumption (fast)
+ * - Token refills: Async background requests to Redis (no blocking)
+ */
+
+const { instance: redisClient } = require('./client');
+const { config } = require('../../../Config');
+const { calculateInterval } = require('./gcra');
+
+// Map of bucket name -> WorkerTokenBucket instance
+const tokenBuckets = new Map();
+
+/**
+ * Per-bucket token bucket for a single worker
+ */
+class WorkerTokenBucket {
+    constructor(bucketName, limitConfig, log) {
+        this.bucketName = bucketName;
+        this.limitConfig = limitConfig;
+        this.log = log;
+
+        // Token buffer configuration
+        this.tokens = 0;
+        this.bufferSize = 50; // Max tokens to hold (5 seconds @ 10 req/s)
+        this.refillThreshold = 20; // Trigger refill when below this
```

Comment:
same here can we have env variables to override to be able to easily tweak settings